cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java

```diff
@@ -127,34 +173,56 @@ public CxxPreprocessor(SquidAstVisitorContext<Grammar> context,
 
     pplineParser = CppParser.create(conf);
 
+    final List<String> configuredDefines = conf.getDefines();
+    Map<String, Macro> configuredMacros = Collections.emptyMap();
+    if (!configuredDefines.isEmpty()) {
+      LOG.debug("parsing configured defines");
+      configuredMacros = parseMacroDefinitions(configuredDefines);
+    }
+
+    // fill CxxPreprocessor.fixedMacros
+    // getMacros() returns CxxPreprocessor.fixedMacros
+    if (getMacros() != fixedMacros) {
+      throw new IllegalStateException("expected fixedMacros as active macros map");
+    }
     try {
       getMacros().setHighPrio(true);
-
-      // parse the configured defines and store into the macro library
-      for (String define : conf.getDefines()) {
-        LOG.debug("parsing external macro: '{}'", define);
-        if (!"".equals(define)) {
-          Macro macro = parseMacroDefinition("#define " + define);
-          LOG.debug("storing external macro: '{}'", macro);
-          getMacros().put(macro.name, macro);
-        }
-      }
-
-      // set standard macros
       getMacros().putAll(Macro.STANDARD_MACROS);
-
-      // parse the configured force includes and store into the macro library
-      for (String include : conf.getForceIncludeFiles()) {
-        LOG.debug("parsing force include: '{}'", include);
-        if (!"".equals(include)) {
-          parseIncludeLine("#include \"" + include + "\"", "sonar." + this.language.getPropertiesKey()
-            + ".forceIncludes", conf.getEncoding());
-        }
-      }
+      getMacros().putAll(configuredMacros);
+      parseForcedIncludes();
     } finally {
       getMacros().setHighPrio(false);
-      ctorInProgress = false;
     }
+
+    // fill CxxPreprocessor.predefinedUnitMacros if relevant
+    final CxxCompilationUnitSettings globalCUSettings = conf.getGlobalCompilationUnitSettings();
+    if (!conf.getCompilationUnitSourceFiles().isEmpty() || (globalCUSettings != null)) {
+      unitMacros = new MapChain<>();
+      if (getMacros() != unitMacros) {
+        throw new IllegalStateException("expected unitMacros as active macros map");
+      }
+      try {
+        getMacros().setHighPrio(true);
+        getMacros().putAll(Macro.UNIT_MACROS);
+        getMacros().putAll(configuredMacros);
+        parseForcedIncludes();
+        predefinedUnitMacros = new HashMap<>(unitMacros.getHighPrioMap());
+      } finally {
+        unitMacros = null;
```

Comment:
@ivangalkin is here `getMacros().setHighPrio(false)` missing?

Reply:
no, in this case we don't have to disable the high-prio mode

1. unitMacros = new MapChain<>(); -> make getMacro() return unitMacros
2. fill the map getMacro() directly and (more important) indirectly through parseForcedIncludes()
3. all precompiled macros for compile units are now in unitMacros.getHighPrioMap(); we copy them into predefinedUnitMacros
4. unitMacros = null; -> destroy the unitMacros; they won't be filled from constructor anymore, no need to toggle the high-prio mode